lib/wit/package.py

```diff
@@ -1,73 +1,149 @@
+        if self.in_root:
+            repo_root = wsroot
+        else:
+            repo_root = wsroot/'.wit'
+            if not repo_root.exists():
+                os.mkdir(str(repo_root))
+
+        self.repo = GitRepo(self.source, self.unresolved_revision, self.name, repo_root)
+
+        # we carefully use Python's boolean expression evalution short-circuiting
+        # to avoid calling has_cmmit if the repo does not exist
+        if (not self.repo.get_path().exists()
+                or non_hash or not self.repo.has_commit(self.unresolved_revision)):
+            if not download:
+                self.repo = None
+                return
+            self.repo.download()
+
+        self.revision = self.repo.get_commit(self.unresolved_revision)
+
+    def is_ancestor(self, other_commit):
+        return self.repo.is_ancestor(other_commit, self.unresolved_revision)
+
+    def find_source(self, repo_paths):
+        for path in repo_paths:
+            tmp_path = str(Path(path) / self.name)
+            if GitRepo.is_git_repo(tmp_path):
+                self.source = tmp_path
+                if self.repo is not None:
+                    self.repo.source = tmp_path
+                return
+
+    def get_dependencies(self):
+        manifest = self.repo.read_manifest_from_commit(self.revision)
+        deps = manifest.dependencies
+        for dep in deps:
+            dep.add_dependent(self)
+        return deps
+
+    def manifest(self):
+        return {
+            'name': self.name,
+            'source': self.source,
+            'commit': self.unresolved_revision,
+        }
+
+    # this is in Package because update_dependency is in Package
+    # it could be confusing to keep the two functions separate
+    def add_dependency(self):
+        """Change the wit-manifest.json to add a dependency."""
+        pass
+
+    def checkout(self, wsroot):
+        """Move to root directory and checkout"""
+        shutil.move(str(self.repo.get_path()), str(wsroot/self.name))
+        self.move_to_root(wsroot)
+        self.repo.revision = self.revision
+        self.repo.checkout()
+
+    def move_to_root(self, wsroot):
+        self.repo.set_wsroot(wsroot)
+        self.repo.name = self.name  # in case we got renamed
+
+    def __repr__(self):
+        return "Pkg({})".format(self.tag())
+
+    def tag(self):
+        return "{}::{}".format(self.name, self.short_revision())
+
+    def get_id(self):
+        return "pkg_"+re.sub(r"([^\w\d])", "_", self.tag())
+
+    def status(self, lock):
+        if lock.contains_package(self.name):
+            if not self.in_root:
+                return "\033[93m(missing)\033[m"
+            elif self.revision != self.repo.get_latest_commit():
+                return "\033[35m(will be checked out to {})\033[m".format(
+                    self.short_revision())
+        else:
+            if not self.in_root:
+                return "\033[92m(will be added to workspace and lockfile)\033[m"
```

Comment:
I don't think it's necessary in this PR, but I think we should factor the colors out into methods. Furthermore, you don't want to use colors when not connected to a terminal (eg. when being piped or redirected). Turning colors off is also often exposed as a command-line option (like `--no-tty` in wake). You can check if connected to a terminal with `sys.stdout.isatty()`

Reply:
Agreed, this would be a good thing to have, and it probably belongs in the logger class somehow.

However, I don't think anyone is parsing wit output, so maybe either:
- we exclude colors from this PR
- OR we fix colors in a different PR

Reply:
It's not just about parsing output, it's also about piping to a log file. Anyway, we can fix it in a follow on PR